pkg/handlers/oauth.go

```diff
@@ -0,0 +1,156 @@
+
+// Package handlers //
+package handlers
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/optimizely/agent/config"
+	"github.com/optimizely/agent/pkg/jwtauth"
+	"github.com/optimizely/agent/pkg/middleware"
+
+	"github.com/go-chi/render"
+)
+
+// ClientCredentials has all info for client credentials
+type ClientCredentials struct {
+	ID     string
+	TTL    time.Duration
+	Secret []byte
+}
+
+// OAuthHandler provides handler for auth
+type OAuthHandler struct {
+	ClientCredentials map[string]ClientCredentials
+	hmacSecret        []byte
+}
+
+type tokenRequest struct {
+	GrantType    string `json:"grant_type"`
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+func renderAccessTokenResponse(w http.ResponseWriter, r *http.Request, accessToken string, expires int64) {
+	// TODO: expires_in should be in seconds, per https://tools.ietf.org/html/rfc6749#section-5.1
+	render.JSON(w, r, tokenResponse{accessToken, "bearer", expires})
+}
+
+// NewOAuthHandler creates new handler for auth
+func NewOAuthHandler(authConfig *config.ServiceAuthConfig) *OAuthHandler {
+
+	clientCredentials := make(map[string]ClientCredentials)
+	// TODO: need to validate all client IDs are unique
+	for _, clientCreds := range authConfig.Clients {
+		clientCredentials[clientCreds.ID] = ClientCredentials{
+			ID:     clientCreds.ID,
+			Secret: []byte(clientCreds.Secret),
+			TTL:    authConfig.TTL,
+		}
+	}
+
+	h := &OAuthHandler{
+		hmacSecret:        []byte(authConfig.HMACSecret),
+		ClientCredentials: clientCredentials,
+	}
+	return h
+}
+
+func (h *OAuthHandler) verifyClientCredentials(r *http.Request) (*ClientCredentials, int, error) {
+	var reqBody tokenRequest
+	err := ParseRequestBody(r, &reqBody)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if reqBody.GrantType == "" {
+		return nil, http.StatusBadRequest, errors.New("grant_type query parameter required")
+	}
+	if reqBody.GrantType != "client_credentials" {
+		return nil, http.StatusBadRequest, fmt.Errorf("unsupported grant_type %v", reqBody.GrantType)
+
+	}
+
+	if reqBody.ClientID == "" {
+		return nil, http.StatusUnauthorized, errors.New("client_id query parameter required")
+	}
+
+	if reqBody.ClientSecret == "" {
+		return nil, http.StatusUnauthorized, errors.New("client_secret query parameter required")
+	}
+	clientCreds, ok := h.ClientCredentials[reqBody.ClientID]
+	// TODO: hash client secret and match secret hash
+	if !ok || !jwtauth.MatchClientSecret(reqBody.ClientSecret, clientCreds.Secret) {
+		return nil, http.StatusForbidden, errors.New("invalid client_id or client_secret")
+	}
+	return &clientCreds, http.StatusOK, nil
+}
+
+// GetAPIAccessToken returns a JWT access token for the API service
+func (h *OAuthHandler) GetAPIAccessToken(w http.ResponseWriter, r *http.Request) {
```

Comment:
Nit, but should these be `Create` as opposed to `Get`? `Get` in a handler sounds like HTTP GET :P